Fix #3041 - Random crashes when you spawn/switch/delete actor

ohlidalp opened the pull request:
See #3041
I fixed one potential crash due to dangling GfxActor pointer (hook holding pointer to DISPOSED actor which doesn't have GfxActor anymore). 

UPDATE: This evolved into signifficant cleanup of the Actor internals. The motivation was to use safer memory handling with added debug checks, but it will also help future extensions.  The original problem turned out to be using non-threadsafe `RefCountingObject<>`s from physics worker threads - see conclusion: https://github.com/RigsOfRods/rigs-of-rods/pull/3042#issuecomment-1557066997

UPDATE2 (2023-05-29): I surrendered and threw a mutex in `RefCountingObject`. Not proud of this at all, but I apparently used it wrong in too many places to look for it now. Technical debt ahoy!

UPDATE3 (2023-06-18): Closing in favor of #3057 (fixes the crashes) and #3059 (refactors the Actor memory management).

ohlidalp committed:
Fixed segfault when removing a linked (hooked) actor.

DISPOSED actor doesn't have GfxActor anymore.

ohlidalp committed:
Eliminated 'ZeroedMemoryAllocator.h'

It's a bad idea to mix `malloc()`/`calloc()` with `delete`, and that's exactly what happened here.
It's especially bad for RefCountingObject which must do `delete this;`, see https://isocpp.org/wiki/faq/freestore-mgmt#delete-this

tritonas00 commented:
```
Thread 1 "RoR" received signal SIGABRT, Aborted.
0x00007ffff649f26c in ?? () from /usr/lib/libc.so.6
(gdb) bt
#0  0x00007ffff649f26c in ?? () from /usr/lib/libc.so.6
#1  0x00007ffff644fa08 in raise () from /usr/lib/libc.so.6
#2  0x00007ffff6438538 in abort () from /usr/lib/libc.so.6
#3  0x00007ffff64392db in ?? () from /usr/lib/libc.so.6
#4  0x00007ffff64a91b7 in ?? () from /usr/lib/libc.so.6
#5  0x00007ffff64ab484 in ?? () from /usr/lib/libc.so.6
#6  0x00007ffff64adcb3 in free () from /usr/lib/libc.so.6
#7  0x0000555555865f9f in RefCountingObject<RoR::SoundScriptInstance>::Release (this=<optimized out>)
    at /home/babis/Downloads/ror-dependencies/rigs-of-rods/source/main/utils/memory/RefCountingObject.h:46
#8  RefCountingObject<RoR::SoundScriptInstance>::Release (this=<optimized out>)
    at /home/babis/Downloads/ror-dependencies/rigs-of-rods/source/main/utils/memory/RefCountingObject.h:40
#9  RefCountingObjectPtr<RoR::SoundScriptInstance>::ReleaseHandle (this=0x5555618b8f48)
    at /home/babis/Downloads/ror-dependencies/rigs-of-rods/source/main/utils/memory/RefCountingObjectPtr.h:211
#10 RefCountingObjectPtr<RoR::SoundScriptInstance>::Set (ref=0x0, this=0x5555618b8f48)
    at /home/babis/Downloads/ror-dependencies/rigs-of-rods/source/main/utils/memory/RefCountingObjectPtr.h:241
#11 RefCountingObjectPtr<RoR::SoundScriptInstance>::operator= (other=..., this=0x5555618b8f48)
    at /home/babis/Downloads/ror-dependencies/rigs-of-rods/source/main/utils/memory/RefCountingObjectPtr.h:231
#12 RoR::Actor::dispose (this=this@entry=0x5555618a5c00) at /home/babis/Downloads/ror-dependencies/rigs-of-rods/source/main/physics/Actor.cpp:117
#13 0x000055555587ada0 in RoR::ActorManager::DeleteActorInternal (this=this@entry=0x555555c72730 <RoR::App::g_game_context+80>, actor=...)
    at /home/babis/Downloads/ror-dependencies/rigs-of-rods/source/main/utils/memory/RefCountingObjectPtr.h:51
#14 0x000055555562c1e9 in RoR::GameContext::DeleteActor (this=<optimized out>, actor=...)
    at /home/babis/Downloads/ror-dependencies/rigs-of-rods/source/main/GameContext.cpp:429
#15 0x00005555555f46a0 in main (argc=<optimized out>, argv=<optimized out>)
    at /home/babis/Downloads/ror-dependencies/rigs-of-rods/source/main/main.cpp:688
    ```

ohlidalp committed:
RefCountingObject destructor fixes

also added RefCountingObject sources to CMakeLists.txt
also cleaned up minor duplicity in ActorManager.cpp

ohlidalp committed:
Fixed AngelScript bindings using `offsetof()` where not allowed.

ProceduralRoads: properties kept using accessor functions. 'road_editor.as' script not changed.
SoundScripts: properties dumped in favor of getter funcs - cleaner anyway because the properties are private in C++. 'demo_script.as' adjusted (with some additional null pointer checks and comments).

ohlidalp committed:
Fixed threading/overlap hazard in event box callbacks.

The script events SE_EVENTBOX_ENTER/EXIT (introduced recently) were called from physics thread pool worker and absolutely unprotected. A new :envelope: MSG_SIM_SCRIPT_EVENT_TRIGGERED was added to resolve the issue.

The classic eventbox handler callbacks (also invoked from physics threadpool worker) used a mutex (`Collisions::m_scriptcallback_mutex`) to prevent overlapping invocation, so the scripts were relatively safe to modify their variables, but nothing prevented them from invoking the game API which was not safe by any means. A new :envelope: MSG_SIM_SCRIPT_CALLBACK_QUEUED was added to resolve this issue.

On top of that, there was a risk of overlapped script callback invocation if an event was triggered from another script. What happened to me was:
 1. I loaded a test script which registered for SE_GENERIC_NEW_TRUCK
 2. I launched an AI vehicle. This works by loading script 'AI.as' which in it's `main()` callback calls `loadTruck()`.
 3. loading the truck triggered the SE_GENERIC_NEW_TRUCK while the AI script was still running, so invoking `eventCallback()` failed with error "script context already active" (asCONTEXT_ACTIVE).

For this reason, I replaced absolutely all event triggers with the async invocation, to avoid future surprises when more of the game is invocable from the script.
Of course this wouldn't have happened if the AI script used `game.pushMessage(MSG_SIM_SPAWN_TRUCK_REQUESTED)`, but that API was introduced later.

ohlidalp committed:
Codechange: `node_t*` -> `NodeNum_t` in AutoPilot

Prepwork for making ar_nodes a std::vector - better diagnostics and flexibility

ohlidalp committed:
Codechange: `node_t*` -> `NodeNum_t` in Airbrake

Prepwork for making ar_nodes a std::vector - better diagnostics and flexibility

ohlidalp committed:
valgrind memcheck: uninitialised value(s) in collisionCorrect()

==129104== Conditional jump or move depends on uninitialised value(s)
==129104==    at 0x4820EE: RoR::Collisions::collisionCorrect(Ogre::Vector<3, float>*, bool) (Collisions.cpp:771)
==129104==    by 0x1FA1D5: RoR::Character::update(float) (Character.cpp:148)
==129104==    by 0x1FB52C: RoR::CharacterFactory::Update(float) (CharacterFactory.cpp:81)
==129104==    by 0x1A404D: main (main.cpp:1010)

ohlidalp committed:
valgrind memcheck: fixed "[RoR|Actor|Warning]  (): Selected module not found: "

(Missing empty string check :|)

==129104== Conditional jump or move depends on uninitialised value(s)
==129104==    at 0x43456C: RoR::ActorSpawner::AddMessage(RoR::ActorSpawner::Message, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (ActorSpawner.cpp:5594)
==129104==    by 0x43606F: RoR::ActorSpawner::ConfigureSections(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::shared_ptr<RigDef::Document>) (ActorSpawner.cpp:101)
==129104==    by 0x42D4A0: RoR::ActorManager::CreateNewActor(RoR::ActorSpawnRequest, std::shared_ptr<RigDef::Document>) (ActorManager.cpp:89)
==129104==    by 0x1DD506: RoR::GameContext::SpawnActor(RoR::ActorSpawnRequest&) (GameContext.cpp:235)
==129104==    by 0x5AFD78: RoR::GameScript::spawnTruckAI(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, Ogre::Vector<3, float>&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, int) (GameScript.cpp:959)
==129104==    by 0x63EC07: X64_CallFunction(unsigned long const*, int, unsigned long (*)(), unsigned long&, bool) (as_callfunc_x64_gcc.cpp:74)

ohlidalp committed:
valgrind memcheck: fixed mem in FlexMesh

CODECHANGES:
raw memory replaced by std::vector<>
also cleaned up node indexing (int -> NodeNum_t)

TRACE:
==129104== Mismatched free() / delete / delete []
==129104==    at 0x484489F: operator delete(void*) (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==129104==    by 0x499388: RoR::FlexMesh::~FlexMesh() (FlexMesh.cpp:240)
==129104==    by 0x499620: RoR::FlexMesh::~FlexMesh() (FlexMesh.cpp:244)
==129104==    by 0x21D9FB: RoR::GfxActor::~GfxActor() (GfxActor.cpp:112)
==129104==    by 0x416DE1: operator() (unique_ptr.h:99)
==129104==    by 0x416DE1: operator() (unique_ptr.h:93)
==129104==    by 0x416DE1: reset (unique_ptr.h:211)
==129104==    by 0x416DE1: reset (unique_ptr.h:509)
==129104==    by 0x416DE1: RoR::Actor::dispose() (Actor.cpp:169)
==129104==    by 0x42B6B2: RoR::ActorManager::DeleteActorInternal(RefCountingObjectPtr<RoR::Actor>) (ActorManager.cpp:880)
==129104==    by 0x1DF298: RoR::GameContext::DeleteActor(RefCountingObjectPtr<RoR::Actor>) (GameContext.cpp:429)
==129104==    by 0x1A5F61: main (main.cpp:688)
==129104==  Address 0x1c9bce90 is 0 bytes inside a block of size 1,600 alloc'd
==129104==    at 0x48432E3: operator new[](unsigned long) (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==129104==    by 0x4998A2: RoR::FlexMesh::FlexMesh(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, RoR::GfxActor*, int, int, int, int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool, float) (FlexMesh.cpp:66)
==129104==    by 0x447C8D: RoR::ActorSpawner::CreateWheelVisuals(unsigned int, unsigned int, unsigned int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool, float) (ActorSpawner.cpp:4964)
==129104==    by 0x454D80: RoR::ActorSpawner::AddWheel(RigDef::Wheel&) (ActorSpawner.cpp:4686)
==129104==    by 0x4635B2: RoR::ActorSpawner::ProcessNewActor(RefCountingObjectPtr<RoR::Actor>, RoR::ActorSpawnRequest, std::shared_ptr<RigDef::Document>) (ActorSpawnerFlow.cpp:147)
==129104==    by 0x42D633: RoR::ActorManager::CreateNewActor(RoR::ActorSpawnRequest, std::shared_ptr<RigDef::Document>) (ActorManager.cpp:90)
==129104==    by 0x1DD506: RoR::GameContext::SpawnActor(RoR::ActorSpawnRequest&) (GameContext.cpp:235)
==129104==    by 0x5AFD78: RoR::GameScript::spawnTruckAI(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, Ogre::Vector<3, float>&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, int) (GameScript.cpp:959)
==129104==    by 0x63EC07: X64_CallFunction(unsigned long const*, int, unsigned long (*)(), unsigned long&, bool) (as_callfunc_x64_gcc.cpp:74)

ohlidalp committed:
Codechange: `node_t*` -> `NodeNum_t` in CameraManager (splinecam)

Prepwork for making ar_nodes a std::vector - better diagnostics and flexibility

ohlidalp committed:
Fixed array out of bounds bug in SoundScriptManager

ohlidalp committed:
Removed limit `MAX_SOUNDSCRIPTS_PER_TRUCK`

Motivation: It's easier to diagnose memory management this way.

ohlidalp committed:
Codechange: `beam_t* -> beamID_t` in `hook_t`

This is prepwork for making ar_beams a std::vector<> for diagnostics and flexibility.

ohlidalp committed:
Codechange: `node_t*` -> `NodeNum_t` in `hook_t`

Prepwork for making ar_nodes a std::vector - better diagnostics and flexibility

ohlidalp committed:
Codechange: `node_t*` -> `NodeNum_t` in `ropable_t`

Prepwork for making ar_nodes a std::vector - better diagnostics and flexibility

ohlidalp committed:
Codechange: `beam_t*/ropable_t* -> BeamID_t/RopableID_t` in `rope_t`

This is prepwork for making ar_beams a std::vector<> - better diagnostics and flexibility.

ohlidalp committed:
Codechange: `beam_t*/ropable_t* -> BeamID_t/RopableID_t` in `tie_t`

This is prepwork for making ar_beams a std::vector<> - better diagnostics and flexibility.

ohlidalp committed:
Codechange: `node_t*` -> `NodeNum_t` in `wheel_t`

Prepwork for making ar_nodes a std::vector - better diagnostics and flexibility

ohlidalp committed:
Codechange: `node_t*` -> `NodeNum_t` in `Actor` (fusedrag)

Prepwork for making ar_nodes a std::vector - better diagnostics and flexibility

ohlidalp committed:
Codechange: `node_t*/AeroEngine**` -> `ActorPtr` in `FlexAirfoil`

this is prepwork for making `ar_nodes` a `std::vector<>` - for more flexibility and easier diagnostics

ohlidalp committed:
Codechange: `node_t*/beam_t*` -> `NodeNum_t/BeamID_t` in `SlideNode`

MOTIVATION:
Prepwork for making ar_nodes/ar_beams a std::vector - better diagnostics and flexibility

CHANGES:
* eliminated struct RailSegment - using just vector<beamID_t>, the looping is handled by simple `bool RailGroup::rg_segments_loop`. The GetPrevSegment()/GetNextSegment() helpers got a little messy, sorry about that.
* deleted most aliases of `getLenTo()` - kept only the static "length to beam" variant and it's non-static slidenode wrapper.
* eliminated `BeamID_t       m_sliding_beam;        //!< current beam sliding on` - redundant, can be fetched from the railgrup

ohlidalp committed:
Codechange: `node_t*` -> `NodeNum_t` in `beam_t`

Prepwork for making ar_nodes a std::vector - better diagnostics and flexibility

ohlidalp committed:
Codechange: `Actor::ar_nodes` is now a `vector<>` :fireworks:

This required a major cleanup across the codebase:
- some objects still referenced the node array separately.
- `ar_num_nodes` find&replaced to `ar_nodes.size()`
- ActorSpawner: all wheel generator funcs had to switch to NodeNum_t

ohlidalp committed:
Codechange: `Actor::ar_beams` is now a `vector<>` :fireworks:

codechanges:
- `ar_num_beams` find&replaced to `ar_beams.size()`
- ActorSpawner: `GetFreeBeam()` now adds new beam.

only-a-ptr commented:
I've piled up a series of "pointer -> index" refactoring commits here. I needed the Actor's node/beam buffers to become `std::vector<>`s to get the advantage of bounds checking and also get rid of `memset()`ing. I wanted to do this anyway for cleaner code and to make Actors editable. 

Interestingly, I can't reproduce the "start 3 AI agoras, then stop, after a few tries -> boom" scenario anymore.

ohlidalp committed:
Codechange: don't use ActorPtr to trigger sounds (threading hazzard)

Sounds are triggered from physics worker threads and RefCountingObject's refcount isn't threadsafe.
There are more hazzards like these...

ohlidalp committed:
Added :envelope: MSG_SIM_ACTOR_LINKING_REQUESTED

Ouch, ouch, ouch, OUCH!!

Well, I investigated this crash scenario: while the autotest.as is running, press J to unpause physics -> insta segfault in physics thread. I narrowed down the investigation to a problem with reference counting of Actors - long story short, the object gets deleted while still in use, resulting in a crash.
It crashed in PointColDetector, and I noticed there's a copy of ActorManager's actor list being made. I quickly started suspecting it to be a hack for multithreading, and sure enough, there was a multithreading hazzard related to it. Hooks were triggered from physics worker threads, so apart from unprotected access to the actor refcount, there was also unprotected access to global linkage table.

I replaced most of the linking calls with the new message, and where not, I commented why it's safe so it shows up in fulltext search. This also cleaned up some typos (using -1 instead of NODENUM_INVALID as param) which didn't actually break anything but were dirty (the param was unused in all the cases).
OVERVIEW:

hookToggle() - invoked from
- CalcTriggers() on physics threadpool worker threads.
- CalcForcesEulerPrepare() on physics background thread.
- VehicleButtons UI - main thread
- SceneMouse::mouseMoved() - main thread
- GameContext::DeleteActor() - processing MSG_SIM_DELETE_ACTOR_REQUESTED

tieToggle() - invoked from:
- ForwardCommands() - physics background thread
- GameContext::DeleteActor() - processing MSG_SIM_DELETE_ACTOR_REQUESTED

ropeToggle() - invoked from:
- ForwardCommands() - physics background thread.

toggleSlideNodeLock() - invoked from:
- RestoreSavedState() - processing MSG_SIM_MODIFY_ACTOR_REQUESTED
- SyncReset() - processing MSG_SIM_MODIFY_ACTOR_REQUESTED, or from NodeBeamUtils UI (main thread)
- vehicle buttons UI (main thread)
- GameContext::SpawnActor() - processing MSG_SIM_ACTOR_SPAWN_REQUESTED

po185021 commented on source/main/physics/ActorManager.cpp:
```diff
@@ -679,11 +679,22 @@ void ActorManager::ForwardCommands(ActorPtr source_actor)
+                    //actor->tieToggle();
+                    ActorLinkingRequest* rq = new ActorLinkingRequest();
+                    rq->alr_type = ActorLinkingRequestType::TIE_ACTION;
+                    rq->alr_actor_instance_id = source_actor->ar_instance_id;
```
oops... bad actor

only-a-ptr replied:
fixed

po185021 commented on source/main/physics/ActorManager.cpp:
```diff
@@ -679,11 +679,22 @@ void ActorManager::ForwardCommands(ActorPtr source_actor)
+                    //actor->ropeToggle(-1);
+                    ActorLinkingRequest* rq = new ActorLinkingRequest();
+                    rq->alr_type = ActorLinkingRequestType::ROPE_ACTION;
+                    rq->alr_actor_instance_id = source_actor->ar_instance_id;
```
ditto

only-a-ptr replied:
fixed

CuriousMike56 commented:
Spamming start/stop on AI gives me:
![image](https://github.com/RigsOfRods/rigs-of-rods/assets/46073351/1a266113-8c6f-4632-b13c-cec45d8a1109)

ohlidalp committed:
Codechange: consolidated auxiliary node data to one place.

data scattered in vector<>s in Actor moved to new struct node_aux_t, see below:
```
struct node_aux_t
{
    int nda_source_id = -1;                   //std::vector<int>         ar_nodes_id;
    std::string nda_source_name;              //std::vector<std::string> ar_nodes_name;
    float nda_minimass;                       //std::vector<float>       ar_minimass;
    float nda_initial_node_mass;              //std::vector<float>        ar_initial_node_masses;
    Ogre::Vector3 nda_initial_node_position;  //std::vector<Ogre::Vector3> ar_initial_node_positions;
    // ...
```

ohlidalp committed:
Codechange: consolidated auxiliary beam data to one place.

```
struct beam_aux_t
{
    float           bma_initial_beam_strength = 0.f; //!< for reset, originally in beam_t
    float           bma_default_beam_deform = 0.f;   //!< for reset, originally in beam_t

    // originally: std::vector<std::pair<float, float>> ar_initial_beam_defaults;
    float           bma_initial_beam_spring = -1.f;
    float           bma_initial_beam_damp = -1.f;
};
```

ohlidalp committed:
Codechange: consolidated camera data to `camera_t`

The following Actor vars were obsoleted by introducing a guideline: if no cameras are defined, generate one with all nodes set to 0 (for backwards compatibility).
```
    NodeNum_t         ar_cameras[0].camera_node_pos            = 0;    //!< Sim attr; ar_camera_node_pos[0]  >= 0 ? ar_camera_node_pos[0]  : 0
    NodeNum_t         ar_main_camera_node_dir            = 0;    //!< Sim attr; ar_camera_node_dir[0]  >= 0 ? ar_camera_node_dir[0]  : 0
    NodeNum_t         ar_main_camera_node_roll           = 0;    //!< Sim attr; ar_camera_node_roll[0] >= 0 ? ar_camera_node_roll[0] : 0
```

The following Actor vars were ransformed to `camera_t` atrrs:
```
    Ogre::Quaternion  ar_main_camera_dir_corr = Ogre::Quaternion::IDENTITY;              //!< Sim attr;
    NodeNum_t         ar_camera_node_pos[MAX_CAMERAS]    = {NODENUM_INVALID};  //!< Physics attr; 'camera' = frame of reference; origin node
    NodeNum_t         ar_camera_node_dir[MAX_CAMERAS]    = {NODENUM_INVALID};  //!< Physics attr; 'camera' = frame of reference; back node
    NodeNum_t         ar_camera_node_roll[MAX_CAMERAS]   = {NODENUM_INVALID};  //!< Physics attr; 'camera' = frame of reference; left node
    bool              ar_camera_node_roll_inv[MAX_CAMERAS] = {false};              //!< Physics attr; 'camera' = frame of reference; indicates roll node is right instead of left
```

ohlidalp committed:
Codechange: `ar_shocks` is now a `vector<>`

ohlidalp committed:
Codechange: `ar_wings` is now a `vector<>`

ohlidalp committed:
Codechange: `ar_wheels` is now a `vector<>`

ohlidalp committed:
Codechange: `ar_rotators` is now a `vector<>`

ohlidalp committed:
Codechange: sorted variables in Actor.h

ohlidalp committed:
Codechange: exposed `ar_linked_actors`

CHANGES:
* Actor.h: exposed `ar_linked_actors`, deleted `getAllLinkedActors()`
* PointColDetector.h/cpp: removed the local copy (it was likely a multithreading hack, but now that we have MSG_SIM_ACTOR_LINKING_REQUESTED it's meaningless)
* The rest is just find&replace (sometimes with corrections related to `auto` keyword)

ohlidalp committed:
Partial fix: do not use `ActorPtr` in `PointColDetector` (not thread safe)

ohlidalp committed:
Completed fix: do not use `ActorPtr` in `PointColDetector` (not thread safe)

ohlidalp committed:
Fix: do not use `ActorPtr` in worker threads (`ActorPtr&` is OK).

only-a-ptr commented:
It should be ready.

Turns out I was really careless when I made `class Actor` a `RefCountingObject<>` - sure, the refcounting was broken before and this helper works flawlessly, except you have to remember not to use it from threads, and I did.

ohlidalp committed:
Hotfix: crash when pressing 'Backspace' (Recovery Mode)

Bug introduced in previous commit - cd85969.
The fix is in RecoveryMode.cpp, the rest is just tidy-up.

ohlidalp committed:
ActorSpawner: separated out `AddHook(NodeNum_t, RigDef&)`

This fixes an array-out-of-bounds if node 0 was also a hooknode - a beam was created between 0 and 1, using nonexistent node 1.

ohlidalp committed:
Workaround: threw a mutex to `RefCountingObject`... not proud.

CuriousMike56 commented:
Windmill Mountain crashes on final "Loading actors" stage (Release)
![image](https://github.com/RigsOfRods/rigs-of-rods/assets/46073351/6447425e-df99-4d22-a261-5bd658c080bf)
Debug result:
![image](https://github.com/RigsOfRods/rigs-of-rods/assets/46073351/46816ab8-d4c0-43e8-a292-675396ffe591)
![image](https://github.com/RigsOfRods/rigs-of-rods/assets/46073351/42e17062-c30d-43ce-98e4-9ddefeae5982)

Debug: F1 track, "Grand prix long" preset, 2 agoras, spamming start/stop eventually gives 
![image](https://github.com/RigsOfRods/rigs-of-rods/assets/46073351/74cfa936-0e64-4b4b-8550-f8d4201ad4ac)
Also got this nasty debug error the first time, no back trace:
![image](https://github.com/RigsOfRods/rigs-of-rods/assets/46073351/91fa7776-f750-4e55-8b86-6879e36e0460)

only-a-ptr commented:
Closing in favor of https://github.com/RigsOfRods/rigs-of-rods/pull/3057 (fixes the crashes) and https://github.com/RigsOfRods/rigs-of-rods/pull/3059 (refactors the Actor memory management).
